truffleruby/src/main/java/org/truffleruby/core/kernel/KernelNodes.java

```diff
@@ -649,7 +646,7 @@ public Object evalBadBinding(DynamicObject source, DynamicObject badBinding, Not
             // TODO CS 14-Apr-15 concat space + code as a rope, otherwise the string will be copied after the rope is converted
             final Source source = Source.newBuilder(space + RopeOperations.decodeRope(code)).name(filename).mimeType(RubyLanguage.MIME_TYPE).build();
 
-            final MaterializedFrame frame = Layouts.BINDING.getFrame(binding);
+            final MaterializedFrame frame = BindingNodes.getExtrasFrame(getContext(), binding);
```

Comment:
We have to ensure the binding has a frame in which the evaluated code could declare new variables. If the binding has an extras frame already then that can be used, but if not one must be created. Otherwise new variables can leak out of this binding into the original bound scope and hence other bindings created from it.